File: python/src/shapes/base/models/circle.py
import math
from typing import final

from shapes.base.commons import BaseShape, Dimensional, Displayable


@final
class Circle(BaseShape, Dimensional, Displayable):
    def __init__(self, radius: float):
        dim_one: float = radius if radius > 0 else 1

        super().__init__(
            [
                {"radius": max(radius, 1)},
                {"diameter": max(radius, 1) * 2},
                {"curvature": 1 / max(radius, 1)},
            ]
        )

        self.radius = dim_one
        self._color = "None"

    @property
    def color(self) -> str:
        return self._color

    @color.setter
    def color(self, value: str) -> None:
        self._color = value

    def print_description(self) -> None:
        print(f"Circle radius: {self.radius}")
        print(f"Circle dimensions: {self._format_dimensions()}")
        print(f"Circle color: {self.color}")
        print(f"Circle area: {self.area()}")
        print(f"Circle perimeter: {self.perimeter()}")
        print(f"Circle is displayable: {self.should_display()}")
        print(f"Circle is a shape: {Circle.validate_shape()}")

    def area(self) -> float:
        return math.pi * self.radius**2

    def perimeter(self) -> float:
        return 2 * math.pi * self.radius

    def _format_dimensions(self) -> str:
        parts: list[str] = [
            f"{key.upper()}: {value:.3f}"
            for dimension in self.dimensions
            for key, value in dimension.items()
        ]

        return " | ".join(parts)

File: ruby/lib/shapes/base/models/circle.rb
# frozen_string_literal: true

require "shapes/base/commons"

class Circle < BaseShape
    include Dimensional
    include Displayable

    attr_accessor :color
    attr_reader :radius

    def initialize(radius)
        dim_one = radius >= 0 ? radius : 1

        super([
            { radius: [radius, 1].max },
            { diameter: [radius, 1].max * 2 },
            { curvature: 1.0 / [radius, 1].max }
        ])

        @radius = dim_one
        @color = "None"
    end

    def print_description
        puts "Circle radius: #{@radius}"
        puts "Circle dimensions: #{format_dimensions}"
        puts "Circle color: #{@color}"
        puts "Circle area: #{area}"
        puts "Circle perimeter: #{perimeter}"
        puts "Circle is displayable: #{should_display}"
        puts "Circle is a shape: #{self.class.validate_shape}"
    end

    def area
        Math::PI * @radius**2
    end

    def perimeter
        2 * Math::PI * @radius
    end

    private

    def format_dimensions
        parts = dimensions.flat_map do |dimension|
            dimension.map { |key, value| "#{key.upcase}: #{format("%.3f", value)}" }
        end

        parts.join(" | ")
    end
end
